### PR TITLE
Geo: extract dateline handling logic from ShapeBuilders

### PR DESCRIPTION
**Pull request by imotov**
Extracts dateline decomposition logic from ShapeBuilder into a separate
utility class that is used on the indexing side. The search side
will be handled as part of another PR at this time we will remove
the decomposition logic from ShapeBuilders as well. This PR also doesn't
change any existing logic including bugs.

Relates to #40908

### CODE REVIEW

**Comment by nknize** [comment]
So it looks like this always assumes spherical geometries and splits across the 180 degree longitude boundary. To support XY cartesian geometries I think we should either:

1. tease out the base geometry building logic into a base class, and move the GIS decomposition logic into a derived class (e.g., `SphericalGeometryIndexer`) so we can extend to a `CartesianGeometryIndexer` in the geo xpack module. 
2. update this class constructor to accept an `isGeo` parameter and only call `decompose` if set to `true`. This way we can build cartesian `Geometry` in core and implement the `XY` geometry building logic in the `toLucenePolygon` method inside `GeometryFieldMapper` of the xpack geo module

I like 2 because I think it makes the building logic more friendly for the REST API?

**Comment by imotov** [comment]
@elasticmachine run elasticsearch-ci/default-distro

@nknize I was thinking about going the first route actually. I would like to make the Indexing logic easily swap-able,  in order to fix the issues that I have stumbled upon this refactoring (such as #43916, #43837, #43826, #43775) in a backward-compatible manner.

File: server/src/main/java/org/elasticsearch/common/geo/GeoJson.java [hunk 1]
```diff
@@ -228,8 +228,12 @@ private XContentBuilder coordinatesToXContent(Polygon polygon) throws IOExceptio
 
     private static Geometry createGeometry(String type, List<Geometry> geometries, CoordinateNode coordinates, Boolean orientation,
                                            boolean defaultOrientation, boolean coerce, DistanceUnit.Distance radius) {
-
-        ShapeType shapeType = ShapeType.forName(type);
+        ShapeType shapeType;
+        if ("bbox".equals(type.toLowerCase(Locale.ROOT))) {
```
Comment (talevy) [inline server/src/main/java/org/elasticsearch/common/geo/GeoJson.java]: would you mind explaining this? I do not see too much test coverage for the "bbox" situation.

File: server/src/test/java/org/elasticsearch/common/geo/GeometryIndexerTests.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+    public void testCircle() {
+        UnsupportedOperationException ex =
+            expectThrows(UnsupportedOperationException.class, () -> indexer.prepareForIndexing(new Circle(1, 2, 3)));
+        assertEquals("CIRCLE geometry is not supported", ex.getMessage());
+    }
+
+    public void testCollection() {
+        assertEquals(GeometryCollection.EMPTY, indexer.prepareForIndexing(GeometryCollection.EMPTY));
+        // TODO: Add other cases
```
Comment (talevy) [inline server/src/test/java/org/elasticsearch/common/geo/GeometryIndexerTests.java]: was this intended to be done in this PR?

File: server/src/main/java/org/elasticsearch/common/geo/GeoJson.java (hunk 1 above)
Comment (imotov) [reply server/src/main/java/org/elasticsearch/common/geo/GeoJson.java]: We support it in ShapeBuilder-based [GeoJson parser](https://github.com/elastic/elasticsearch/blob/df4a8147e04b4c1424414170873b2ec008905a05/server/src/main/java/org/elasticsearch/common/geo/GeoJson.java#L232) and test it [here](https://github.com/elastic/elasticsearch/blob/5fb55f62be68add6530cc883e317cb764bf4ad1e/server/src/test/java/org/elasticsearch/search/geo/GeoShapeIntegrationIT.java#L146), but I will add an additional unit-level test for it to GeoJson parser.

File: server/src/test/java/org/elasticsearch/common/geo/GeometryIndexerTests.java (hunk 2 above)
Comment (imotov) [reply server/src/test/java/org/elasticsearch/common/geo/GeometryIndexerTests.java]: Yes! Thanks for catching it!

**Review by talevy — APPROVED** [review]
LGTM I can't wait for the follow-up PR that removes all the other stuff!

**Comment by nknize** [comment]
@imotov Just a quick follow up from our conversation: are you still planning on moving `GeometryIndexer` out of `o.e.common.geo` since it's only used in `GeoShapeFieldMapper`? I assumed we would do it in this PR. I think it's fine to move it as package private to `o.e.index.mapper` since it's never used outside of the Mapping?

**Comment by imotov** [comment]
>  think it's fine to move it as package private to o.e.index.mapper since it's never used outside of the Mapping?

I don't think I can do that. It will probably be needed in QueryBuilders.

**Review by nknize — APPROVED** [review]
> It will probably be needed in QueryBuilders.
Okay, let's do it in a separate PR then when we refactor the queries.

**Pull request merged**